### PR TITLE
REF: add internal mechanics to separate index/columns sparsification in Styler

### PR DESCRIPTION
**Pull request by attack68**
This is not a user facing change.

It adds only the necessary code to permit separate control of index/column MultIndex sparsification. Essentially it removes the `get_option("display.multi_sparse")` from a low level method to a higher level method where it can be overwritten by user-args later.

partly addresses #41142

### CODE REVIEW

File: pandas/io/formats/style_render.py [hunk 1]
```diff
@@ -520,14 +532,23 @@ def _element(
...
     """
     Given an index, find the level length for each element.
 
-    Optional argument is a list of index positions which
-    should not be visible.
-
-    Result is a dictionary of (level, initial_position): span
+    Parameters
+    ----------
+        index : Index
```
Comment (jreback) [inline pandas/io/formats/style_render.py:541]: is this indented?

File: pandas/io/formats/style_render.py [hunk 2]
```diff
@@ -520,14 +532,23 @@ def _element(
...
+            Index or columns to determine lengths of each element
+        sparsify : bool
+            Whether to hide or show each distinct element in a MultiIndex
+        hidden_element : list
+            Index positions of elements hidden from display in the index affecting
+            length
+    Returns
```
Comment (jreback) [inline pandas/io/formats/style_render.py:548]: blank line before this

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm doc comments.

File: pandas/io/formats/style_render.py (hunk 1 above)
Comment (attack68) [reply pandas/io/formats/style_render.py]: fixed

File: pandas/io/formats/style_render.py (hunk 2 above)
Comment (attack68) [reply pandas/io/formats/style_render.py:549]: fixed

File: pandas/io/formats/style_render.py [hunk 3]
```diff
@@ -132,11 +132,16 @@ def _compute(self):
             r = func(self)(*args, **kwargs)
         return r
 
-    def _translate(self):
+    def _translate(self, sparsify_index=None, sparsify_cols=None):
```
Comment (jreback) [inline pandas/io/formats/style_render.py:135]: can you type input & output here

Comment (jreback) [reply pandas/io/formats/style_render.py:135]: docs-strings as well

File: pandas/io/formats/style_render.py [hunk 4]
```diff
@@ -520,14 +532,24 @@ def _element(
     }
 
 
-def _get_level_lengths(index, hidden_elements=None):
+def _get_level_lengths(index, sparsify, hidden_elements=None):
```
Comment (jreback) [inline pandas/io/formats/style_render.py:535]: can you type

**Review by jreback — APPROVED** [review]
looks fine. happy to merge and do typing on followups or here.

File: pandas/io/formats/style_render.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style_render.py]: typed.

File: pandas/io/formats/style_render.py (hunk 3 above)
Comment (attack68) [reply pandas/io/formats/style_render.py]: typed and added all doc strings

**Comment by attack68** [comment]
@jreback ping this is green and your requested changes made.

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**